### PR TITLE
redolog: add DeleteAllLogs api

### PR DESCRIPTION
**Pull request by ben1009**
### What problem does this PR solve? 
close https://github.com/pingcap/ticdc/issues/3347
### What is changed and how it works?

- add redolog  DeleteAllLogs api at storage to support delete all redo logs when delete related changedfeed
- remove local file first when apply logs if s3 enabled

### Check List 

Tests 

 - Unit test

Code changes

 - Has interface methods change

Side effects

Related changes

### Release note 
```release-note
 None
```

### CODE REVIEW

File: cdc/redo/reader/reader_test.go [hunk 1]
```diff
@@ -38,6 +38,23 @@ func TestNewLogReader(t *testing.T) {
...
+	//ctx, cancel := context.WithCancel(context.Background())
+	//cancel()
+	//_, err = NewLogReader(ctx, &LogReaderConfig{
+	//	S3Storage: true,
+	//	Dir:       dir,
+	//	S3URI:     *s3URI,
+	//})
+	//require.Regexp(t, ".*ErrRedoDownloadFailed*.", err)
+	//_, err = os.Stat(dir)
+	//require.True(t, os.IsNotExist(err))
```
Comment (amyangfei) [inline cdc/redo/reader/reader_test.go:57]: Why are these codes commented?

File: cdc/redo/writer/file.go [hunk 2]
```diff
@@ -177,15 +177,10 @@ func (w *Writer) runFlushToDisk(ctx context.Context, flushIntervalInMs int64) {
 			return
 		}
 
-		select {
-		case <-ctx.Done():
```
Comment (amyangfei) [inline cdc/redo/writer/file.go:181]: Why is the context check removed here?

File: cdc/redo/writer/writer.go [hunk 3]
```diff
@@ -212,16 +225,12 @@ func (l *LogWriter) runGC(ctx context.Context) {
 			return
 		}
 
-		select {
-		case <-ctx.Done():
```
Comment (amyangfei) [inline cdc/redo/writer/writer.go:216]: Ditto, why is context check removed?

**Comment by ben1009** [comment]
lgtm about owner call DeleteAllLogs part

File: cdc/redo/manager.go [hunk 4]
```diff
@@ -302,7 +305,11 @@ func (m *ManagerImpl) RemoveTable(tableID model.TableID) {
 	} else {
 		log.Warn("remove a table not maintained in redo log manager", zap.Int64("table-id", tableID))
 	}
-	// TODO: send remove table command to redo log writer
```
Comment (amyangfei) [inline cdc/redo/manager.go:305]: Redo log applier will sort redo logs, so the writer doesn't need to care about table scheduling.

File: cdc/redo/writer/writer.go [hunk 5]
```diff
@@ -99,72 +105,79 @@ type LogWriter struct {
...
-			return
+	initLock.Lock()
+	defer initLock.Unlock()
+
+	if v, ok := logWriters[cfg.ChangeFeedID]; ok {
+		// if cfg changed or already closed need create a new LogWriter
+		if cfg.String() == v.cfg.String() && !v.isStopped() {
+			return v, nil
 		}
+	}
```
Comment (amyangfei) [inline cdc/redo/writer/writer.go:121]: In which scenario will cfg changed and create a new redo log writer? I can think of one
`pause changefeed -> changefeed update config -> resume changefeed`, is there any other scenario?

File: cdc/redo/writer/writer.go [hunk 6]
```diff
@@ -99,72 +105,79 @@ type LogWriter struct {
...
+			return nil, err
 		}
-
-		go logWriter.runGC(ctx)
-	})
-	if errInit != nil {
-		return nil, errInit
 	}
+	logWriters[cfg.ChangeFeedID] = logWriter
+	go logWriter.runGC(ctx)
```
Comment (amyangfei) [inline cdc/redo/writer/writer.go:180]: If a changefeed is paused, the `ctx` will be cancelled. After the changefeed is resumed, will the `runGC` goroutine run again?

File: cdc/redo/writer/writer.go (hunk 5 above)
Comment (ben1009) [reply cdc/redo/writer/writer.go:121]: yes, like change file size, flush interval, or delete changefeed

File: cdc/redo/writer/file.go [hunk 7]
```diff
@@ -162,7 +155,7 @@ func NewWriter(ctx context.Context, cfg *FileWriterConfig, opts ...Option) (*Wri
 		storage:   s3storage,
 	}
 
-	w.state.Store(started)
+	w.state.Store(true)
```
Comment (amyangfei) [inline cdc/redo/writer/file.go:158]: What about renaming `state` to a more explicit name such as `isRunning`

**Review by hi-rustin — APPROVED** [review]

**Review by amyangfei — APPROVED** [review]

**Pull request merged**